src/services/DomainModel.ts

```diff
@@ -0,0 +1,125 @@
+// Do not assume you can reach every node, sensor, or reading starting from a
+// gateway you're looking at. Be mindful that a QueryFilter is probably in
+// effect which has limited the extent to which we have populated all the
+// objects into memory. For example, if you're working with recent data from the
+// 'Black Pearl' node you might have info about the Gateway, but just the one
+// node, and a few days of readings. You'll find `Unknown`s at the points in the
+// tree that we don't care about in the context of an active Query.
+
+export const unknown = Symbol("Unknown");
+export type Unknown = typeof unknown;
+
+// These `...ID` types prevent `let gatewayID = getNodeID()` which would sadly
+// compile if each were simply string. Is there an easier way? Is it worth it?
+export type GatewayID = { type: "GatewayID"; id: string };
```

Comment:
I like the intent, but does this mean that suddenly types are present at runtime?

Ideally, rather than working with IDs, we should be using objects and object references. (A getter for a known type.) That way there is no need for making synthetic IDs typesafe.

Reply:
I like where you're headed with this, but what I haven't figured out is this: 

We can work with objects and js object references once things are in javascript-land. But what about while things are moving though HTTP? (e.g. a request to GET /gateway/dev:5555)

I don't want to have to think about HTTP implementation details while thinking about the domain-model, but people do know their car has a VIN because they need to use it to prove ownership, register it, get insurance, find replacement parts, etc. It's part of the UI of a car for better or worse. Do you think the actual domain objects should have IDs but hidden as private attributes and not part of the DomainModel type system?

Reply:
Figured out!